velox/exec/Window.cpp

```diff
@@ -282,6 +327,117 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
+        VELOX_USER_FAIL(
+            "{} in frame bound at index {} must be at least 1", offsets[i], i);
+      }
+    }
+
+    for (auto i = 0; i < numRows; i++) {
+      rawFrameBounds[i] = (startRow + i) +
+          (isKPreceding ? -vector_size_t(offsets[i])
+                        : vector_size_t(offsets[i])) -
+          firstPartitionRow;
+    }
+  }
+}
+
+void Window::updateFrameBounds(
+    const vector_size_t& i,
+    const bool& isStartBound,
+    const vector_size_t& startRow,
+    const vector_size_t& numRows) {
+  auto firstPartitionRow = partitionStartRows_[currentPartition_];
+  auto lastPartitionRow = partitionStartRows_[currentPartition_ + 1] - 1;
+  auto rawPeerStarts = peerStartBuffer_->asMutable<vector_size_t>();
+  auto rawPeerEnds = peerEndBuffer_->asMutable<vector_size_t>();
+  auto rawFrameBounds = isStartBound
+      ? frameStartBuffers_[i]->asMutable<vector_size_t>()
+      : frameEndBuffers_[i]->asMutable<vector_size_t>();
+  auto type = windowFrames_[i].type;
+  auto boundType =
+      isStartBound ? windowFrames_[i].startType : windowFrames_[i].endType;
+  auto frameArg = isStartBound ? windowFrames_[i].start : windowFrames_[i].end;
+
+  switch (boundType) {
+    case core::WindowNode::BoundType::kUnboundedPreceding:
+      std::memset(rawFrameBounds, 0, numRows * sizeof(vector_size_t));
+      break;
+    case core::WindowNode::BoundType::kUnboundedFollowing:
+      std::fill_n(
+          rawFrameBounds, numRows, lastPartitionRow - firstPartitionRow);
+      break;
+    case core::WindowNode::BoundType::kCurrentRow: {
+      if (type == core::WindowNode::WindowType::kRange) {
+        const vector_size_t* rawPeerBuffer =
+            isStartBound ? rawPeerStarts : rawPeerEnds;
+        std::copy(rawPeerBuffer, rawPeerBuffer + numRows, rawFrameBounds);
+      } else {
+        // Fills the frameBound buffer with increasing value of row indices
+        // (corresponding to CURRENT ROW) from the startRow of the current
+        // output buffer. The startRow has to be adjusted relative to the
+        // partition start row.
+        std::iota(
+            rawFrameBounds,
+            rawFrameBounds + numRows,
+            startRow - firstPartitionRow);
+      }
+      break;
+    }
+    case core::WindowNode::BoundType::kPreceding: {
+      updateKRowsFrameBounds(
+          true, frameArg.value(), startRow, numRows, rawFrameBounds);
+      break;
+    }
+    case core::WindowNode::BoundType::kFollowing: {
+      updateKRowsFrameBounds(
+          false, frameArg.value(), startRow, numRows, rawFrameBounds);
+      break;
+    }
+    default:
+      VELOX_USER_FAIL("Invalid frame bound type");
+  }
+}
+
+namespace {
+void fixOutOfBoundFrames(
+    const vector_size_t& lastRow,
+    const vector_size_t& numRows,
+    vector_size_t* frameStart,
+    vector_size_t* frameEnd) {
+  for (auto i = frameStart, j = frameEnd; i < frameStart + numRows; i++, j++) {
+    if (*i <= *j) {
+      *i = std::max(*i, 0);
```

Comment:
Add a comment "Clamp frameStart and frameEnd to the first and last row of the partition if required"